### PR TITLE
Add clock from apimachinery 

### PR DESCRIPTION
**Pull request by dashpole**
My end goal is to use the clock in cAdvisor.  Because of the way that staging works in kubernetes/kubernetes, the import loop kubernetes/apimachinery -> google/cadvisor -> kubernetes/kubernetes is not allowed.
I would copy the clock code to cadvisor, but cadvisor is currently google-owned, and I dont believe copying it from CNCF to google is permitted.

cc @tallclair

### CODE REVIEW

**Comment by dashpole** [comment]
I brought over history from apimachinery

**Comment by mengqiy** [comment]
I think this LG.
It seems this code is copied from `k8s.io/apimachinery` instead of `/staging` dir. So each commit has [`k8s-publish-bot`](https://github.com/k8s-publish-robot)'s commit. IMO it will be better to not have the bot commits.

And that publish bot has been down for weeks, so please double check what you copied is synced with the latest code in `/staging`.

@apelisse WDYT?

**Comment by dashpole** [comment]
oh, I should copy directly from staging

**Comment by dashpole** [comment]
give me a minute

**Comment by apelisse** [comment]
Good catch Mengqi, I would never have noticed that

**Comment by dashpole** [comment]
commits are now from kubernetes/kubernetes

**Comment by mengqiy** [comment]
/lgtm
Waiting for travis to be green.

**Comment by apelisse** [comment]
Yeah, I'd be happy to see more examples and better documentation, but won't delay any work on that of course.

**Comment by apelisse** [comment]
One thing that I'd actually like to see change now (sooner rather than later) is:
- Move testing code to its own testing package, so that it's not included in production code.

**Comment by dashpole** [comment]
when you import with godep, it doesnt bring in testing code, right?

**Comment by dashpole** [comment]
or you mean just splitting the mock and real implementations?

**Comment by apelisse** [comment]
> when you import with godep, it doesnt bring in testing code, right?

But `FakeClock` and `IntervalClock` live in `clock.go`, next to `RealClock`.

**Comment by dashpole** [comment]
yes, agreed

File: clock/clock.go [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+func (RealClock) Tick(d time.Duration) <-chan time.Time {
+	return time.Tick(d)
+}
+
+func (RealClock) Sleep(d time.Duration) {
+	time.Sleep(d)
+}
+
+// FakeClock implements Clock, but returns an arbitrary time.
+type FakeClock struct {
```
Comment (thockin) [inline clock/clock.go]: We should put fakes and test-related things in sub-packages, typically called "testing", so they don't get linked into prod binaries.

@dashpole

Comment (dashpole) [reply clock/clock.go]: done https://github.com/kubernetes/utils/pull/25/

Comment (apelisse) [reply clock/clock.go]: Yeah, we noticed that: #25

**Pull request merged**